plugins/PostProcessingPlugin/scripts/PurgeLinesAndUnload.py

```diff
@@ -0,0 +1,752 @@
+
+        elif bed_shape == "elliptic" and origin_at_center:
+            radius = machine_width / 2
+            offset_sin = round(2**.5 / 2 * radius, 2)
+            if purge_end_loc == "LR":
+                if goto_str == "LtFrt":
+                    move_str += f"G0 F{travel_speed} X-{offset_sin} Z2 ; Move\nG0 Y-{offset_sin} Z2 ; Move to start\n"
+                elif goto_str == "LtBk":
+                    move_str += f"G2 X0 Y{offset_sin} I{offset_sin} J{offset_sin} ; Move around to start\n"
+                elif goto_str == "RtFrt":
+                    move_str += f"G0 F{travel_speed} X{offset_sin} Z2 ; Ortho move\nG0 Y-{offset_sin} Z2 ; Ortho move\n"
+                elif goto_str == "RtBk":
+                    move_str += f"G0 F{travel_speed} X{offset_sin} Z2 ; Ortho move\nG0 Y{offset_sin} Z2 ; Ortho move\n"
+            elif purge_end_loc == "RR":
+                if goto_str == "LtFrt":
+                    move_str += f"G0 F{travel_speed} X-{offset_sin} Z2 ; Move\nG0 Y-{offset_sin} Z2 ; Move to start\n"
+                elif goto_str == "LtBk":
+                    move_str += f"G0 F{travel_speed} X-{offset_sin} Z2 ; Ortho move\nG0 Y{offset_sin} Z2 ; Ortho move\n"
+                elif goto_str == "RtFrt":
+                    move_str += f"G0 F{travel_speed} X{offset_sin} Z2 ; Ortho move\nG0 Y-{offset_sin} Z2 ; Ortho move\n"
+                elif goto_str == "RtBk":
+                    move_str += f"G0 F{travel_speed} X{offset_sin} Z2 ; Ortho move\nG0 Y{offset_sin} Z2 ; Ortho move\n"
+            elif purge_end_loc == "LF":
+                if goto_str == "LtFrt":
+                    move_str += f"G0 F{travel_speed} X-{offset_sin} Z2 ; Move\nG0 Y-{offset_sin} Z2 ; Move to start\n"
+                elif goto_str == "LtBk":
+                    move_str += f"G0 F{travel_speed} X-{offset_sin} Z2 ; Ortho move\nG0 Y{offset_sin} Z2 ; Ortho move\n"
+                elif goto_str == "RtFrt":
+                    move_str += f"G0 F{travel_speed} X{offset_sin} Z2 ; Ortho move\nG0 Y-{offset_sin} Z2 ; Ortho move\n"
+                elif goto_str == "RtBk":
+                    move_str += f"G0 F{travel_speed} X{offset_sin} Z2 ; Ortho move\nG0 Y{offset_sin} Z2 ; Ortho move\n"
+        move_str += ";---------------------[End of layer start travels]"
+        # Add the move_str to the end of the StartUp section and move 'LAYER_COUNT' to the end.
+        startup = data[1].split("\n")
+        for index, line in enumerate(startup):
+            if "LAYER_COUNT" in line:
+                lay_count = startup.pop(index) + "\n"
+                break
+        move_str = self._format_string(move_str)
+        if move_str.startswith("\n"):
+            move_str = move_str[1:]
+        startup.append(move_str)
+        startup.append(lay_count)
+        data[1] = "\n".join(startup)
+        # Remove any double spaced lines
+        data[1] = data[1].replace("\n\n", "\n")
+        return
```

Comment:
same here (return data)

Reply:
It might be a post-processor thing but sometimes (maybe all the time) a "return data" will send the "data" list back to Cura assuming that all the work is done.  The script ends when that happens even though there might be more to do.

My understanding is that "data" is global so when the value is changed as in "data[1] = data[1].replace("\n\n", "\n")" then that is the value of data throughout the script and "return data" would be redundant.  That's why the only "return data" in the script was in the Execute function.

I'm self taught so my coding is decidedly un-pythonic.  It's more of a mish-mash of Python, Visual Basic, the Excel 4.0 macro language, and Fortran I.  I mention Fortran I because it was the only actual programming class I've ever taken.  That was in 1969.  I know where I want to go with the code, but the way I get there is rarely a straight line.

If you ever look at "Advanced Cooling Fan Control" you will be horrified.  Remco certainly was when he looked at it.  I think at some point he threw up his hands in disgust and just ran with it since the code worked.  It's just ugly.  There have been exactly zero bug reports about it.

Make whatever changes you feel are necessary because most of the time I don't really understand what you are asking.  This script is one I keep active for every print so I thought I'd pass it along.  It's handy.  If you want to throw it out be my guest.  It won't hurt my feelings.

Reply:
I learned coding with Pascal so I have the same approach of first make it work and then worry about it not being pretty. Then many years later I had to make a 4D scanner and needed to process videos from 100 cameras into frames and sync them together through an audio trigger, so I started with python. So I see coding as a tool, if it gets the job done, that's all that matters :P (but it's personal opinion, it gets a bit difficult with more people in a team and that's why you get a coding style and such).

The return data thing was a question cause I attempted to run the script but it didn't generate any gcode commands for me, so that seemed like a usual suspect. But if you have it running then it might be something else I'm missing.

In some cases, I can't write anything better either, it's just a lot of checks and a lot of conditions, so the suggestions you see are mostly tips and tricks, they might help for something in future or they might not (making it harder to understand) that's why I add them as suggestions :) . I'm sure the devs will write helper functions and do some fancy stuff. But my questions or suggestions are mainly me trying to understand the problem or issues that you had and the process that you followed to resolve it which resulted in the code. 

TLDR:
I'll poke around some more, see if I can get it working on my machine.